### PR TITLE
feat(api,vm) Improve the `vm::ImportInitializerFuncPtr` type.

### PR DESCRIPTION
**Pull request by Hywan**
# Description

This PR is a little bit of code cleanup. Let's higlight the most important commits:

* https://github.com/wasmerio/wasmer/commit/48a5cf0b55cd4592e365fc96eed6f3fdb3913480 that re-use the `vm::ImportInitializerFuncPtr` type inside `api::externals::function`. It makes type-checking useful, and it makes code navigation much easier,
* https://github.com/wasmerio/wasmer/commit/0960f0fa7dc41708beb21a6aeeadcd98e15b1dbf that adds a new optional `ResultErr` parameter to `ImportInitializerFuncPtr`. It simplifies the intend when transmuting from `&ImportInitializerFuncPtr` to `&ImportInitializerFuncPtr<Err>` in `vm::instance`.

The diff is annotated with comments to guide the reviewer.

# Review

- ~[ ] Add a short description of the the change to the CHANGELOG.md file~ not necessary I guess?

### CODE REVIEW

File: lib/api/src/externals/function.rs [hunk 1]
```diff
@@ -73,24 +74,22 @@ fn build_export_function_metadata<Env>(
...
-) -> (*mut std::ffi::c_void, ExportFunctionMetadata)
+) -> (*mut c_void, ExportFunctionMetadata)
 where
     Env: Clone + Sized + 'static + Send + Sync,
 {
     let import_init_function_ptr = Some(unsafe {
-        std::mem::transmute::<fn(_, _) -> Result<(), _>, fn(_, _) -> Result<(), _>>(
-            import_init_function_ptr,
-        )
+        std::mem::transmute::<_, ImportInitializerFuncPtr>(import_init_function_ptr)
```
Comment (Hywan) [inline lib/api/src/externals/function.rs:82]: Transmuting here is simplified and clearer, _and_ most importantly, it uses an existing type alias, which prevents future bugs, and improves code navigation.

File: lib/c-api/src/wasm_c_api/externals/function.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ pub type wasm_func_callback_t = unsafe extern "C" fn(
 
 #[allow(non_camel_case_types)]
 pub type wasm_func_callback_with_env_t = unsafe extern "C" fn(
-    *mut c_void,
+    env: *mut c_void,
```
Comment (Hywan) [inline lib/c-api/src/wasm_c_api/externals/function.rs:26]: Other arguments have a label. Now, this one has a label too :-).

File: lib/c-api/src/wasm_c_api/externals/function.rs [hunk 3]
```diff
@@ -115,18 +115,18 @@ pub unsafe extern "C" fn wasm_func_new_with_env(
...
                 if !self.env.is_null() {
-                    unsafe { (finalizer)(self.env as _) }
+                    unsafe { (env_finalizer)(self.env as _) }
                 }
             }
         }
     }
 
-    let inner_callback = move |env: &WrapperEnv, args: &[Val]| -> Result<Vec<Val>, RuntimeError> {
+    let trampoline = move |env: &WrapperEnv, args: &[Val]| -> Result<Vec<Val>, RuntimeError> {
```
Comment (Hywan) [inline lib/c-api/src/wasm_c_api/externals/function.rs:129]: Let's call a cat a cat.

File: lib/vm/src/instance/mod.rs [hunk 4]
```diff
@@ -47,8 +47,8 @@ use wasmer_types::{
 
 /// The function pointer to call with data and an [`Instance`] pointer to
 /// finish initializing the host env.
-pub type ImportInitializerFuncPtr =
-    fn(*mut std::ffi::c_void, *const std::ffi::c_void) -> Result<(), *mut std::ffi::c_void>;
+pub type ImportInitializerFuncPtr<ResultErr = *mut ffi::c_void> =
+    fn(*mut ffi::c_void, *const ffi::c_void) -> Result<(), ResultErr>;
```
Comment (Hywan) [inline lib/vm/src/instance/mod.rs:51]: The new optional `ResultErr` parameter. It's useful below.

File: lib/vm/src/instance/mod.rs [hunk 5]
```diff
@@ -1146,7 +1146,7 @@ impl InstanceHandle {
                         // transmute our function pointer into one with the correct error type
                         let f = mem::transmute::<
                             &ImportInitializerFuncPtr,
-                            &fn(*mut ffi::c_void, *const ffi::c_void) -> Result<(), Err>,
+                            &ImportInitializerFuncPtr<Err>,
```
Comment (Hywan) [inline lib/vm/src/instance/mod.rs:1149]: The new optional `ResultErr` parameter of `ImportInitializerFuncPtr` is useful here. Less code repetition, more clarity, it prevents future bug.

**Comment by Hywan** [comment]
bors try

**Comment by Hywan** [comment]
Tests are failing due to an error in the CI, it's not related to this patch.

**Review by syrusakbary — APPROVED** [review]
This looks great to me. Although I’d like @MarkMcCaskey to do the full review

**Comment by syrusakbary** [comment]
Merging manually as I want to speed up the wasmer-go release

**Pull request merged**